### PR TITLE
Feature TreeObservable

### PR DESCRIPTION
**Pull request by sebbert**
~~Requires #490~~ (now merged)

This PR contains:
- [ ] Changelog
- [ ] Documentation
- [x] Tests

### CODE REVIEW

File: Source/Fuse.Reactive.JavaScript/ArrayMirror.uno [hunk 1]
```diff
@@ -27,6 +27,9 @@ namespace Fuse.Reactive
 
 		public override void Unsubscribe()
 		{
+			if (_isDisposed) return;
+			_isDisposed = true;
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/ArrayMirror.uno]: This feels a bit like abuse. Unsubscribe and Dispose isn't quite the same... Perhaps they should be, though?

File: Source/Fuse.Reactive.JavaScript/Tests/UX/TreeObservable.Set.ux [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<Panel ux:Class="UX.TreeObservable.Set">
+	<JavaScript>
+		var TreeObservable = require("FuseJS/TreeObservable");
+		function Tree() {}
+		Tree.prototype = Object.create(TreeObservable.prototype);
+		let tree = new Tree();
+
+		tree.__fuse_id = 0;
```
Comment (mortoray) [inline Source/Fuse.Reactive.JavaScript/Tests/UX/TreeObservable.Set.ux]: What's with the explicit setting of `__fuse_id` ? Isn't this automatic, or is it something the model differ does later?

File: Source/Fuse.Reactive.JavaScript/ObjectMirror.uno [hunk 3]
```diff
@@ -28,6 +28,9 @@ namespace Fuse.Reactive
 
 		public override void Unsubscribe()
 		{
+			if (_isDisposed) return;
+			_isDisposed = true;
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/ObjectMirror.uno]: Same comment as above, this feels like an abuse...

File: Source/Fuse.Reactive.JavaScript/ValueMirror.uno [hunk 4]
```diff
@@ -22,6 +22,7 @@ namespace Fuse.Reactive
 		readonly object _raw;
 		public object Raw { get { return _raw; } }
 		public virtual object ReflectedRaw { get { return _raw; } }
+		protected bool _isDisposed = false;
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/ValueMirror.uno]: I don't think this is the right place to introduce this. I think it's better to introduce these in ObjectMirror and ArrayMirror. And then we can maybe even name it a bit better...

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+					var newValuesJs = ctx.NewArray(nv);
+
+					var replaceAllFn = (Scripting.Function) ctx.Evaluate("replaceAll",
+						"(function(array, values) {" +
+							"if('__fuse_replaceAll' in array) array.__fuse_replaceAll(values);" +
+							"else {"+
+								"array.length = 0;"+
+								"Array.prototype.push.apply(array, values);"+
+							"}"+
+						"})");
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: Looks like this one is constant, do we really need to re-create it every time?

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeObject.uno [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+					else
+					{
+						_obj[_key] = _value;
+					}
+				}
+			}
+
+			bool IPropertySubscription.TrySetExclusive(string key, object newValue)
+			{
+				var t = SubscriptionSubject as TreeObject;
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeObject.uno]: Please do
```uno
var t = (TreeObject)SubscriptionSubject;
```

This is clearer. We're going to be dereferencing `t`  on the next statement anyway, so it needs to be non-null. Doing a proper cast gives us a better error-message if this was every wrong.

File: Source/Fuse.Reactive/ObserverMap.uno [hunk 7]
```diff
@@ -54,14 +54,16 @@ namespace Fuse.Reactive
 		{
 			Detach();
 			_observable = obs;
-			_slave = slave;
```
Comment (mortoray) [inline Source/Fuse.Reactive/ObserverMap.uno]: This bit is likely to conflict with https://github.com/fusetools/fuselibs-public/pull/487
That PR also fixes this issue, in the same way, but with other changes as well.  Maybe we can remove the changes to ObserverMap from this PR and get that other PR merged to 1.3, then merged to master first? Or just deal with conflict later?

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+			_items.Clear();
+
+			for(var i = 0; i < newValues.Length; ++i)
+			{
+				_items.Add(newValues[i]);
+			}
+
+			var sub = Subscribers as ArraySubscription;
+			if(sub != null)
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: space between `if` and opening paren.

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+			public void OnRemoveAt(int index)
+			{
+				_observer.OnRemoveAt(index);
+				var next = Next as ArraySubscription;
+				if (next != null) next.OnRemoveAt(index);
+			}
+
+			public void OnReplaceAll(IArray values, ArraySubscription exclude)
+			{
+				if(this != exclude) _observer.OnNewAll(values);
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: space between `if` and opening paren.

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+				var next = Next as ArraySubscription;
+				if(next != null) next.OnReplaceAll(values, exclude);
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: space between `if` and opening paren.

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+					_target = target;
+					_newValues = newValues;
+				}
+
+				public void Perform()
+				{
+					var ctx = _worker.Context;
+
+					var nv = new object[_newValues.Length];
+					for(var i = 0; i < _newValues.Length; ++i) {
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: space between `for` and opening paren.

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+					{
+						return _values[index];
+					}
+				}
+			}
+		}
+
+		internal void ReplaceAll(IArray newValues, ArraySubscription exclude)
+		{
+			for(var i = 0; i < _items.Count; ++i)
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: space between `for` and opening paren.

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+			{
+				ValueMirror.Unsubscribe(_items[i]);
+			}
+
+			_items.Clear();
+
+			for(var i = 0; i < newValues.Length; ++i)
```
Comment (kusma) [inline Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: space between `for` and opening paren.

File: Source/Fuse.Reactive.JavaScript/Tests/UX/TreeObservable.Set.ux (hunk 2 above)
Comment (sebbert) [reply Source/Fuse.Reactive.JavaScript/Tests/UX/TreeObservable.Set.ux]: These are automatically generated by Model system, yes. It's an ugly workaround for the fact that `Scripting.Object` doesn't guarantee object equality/identity on the Uno side.

File: Source/Fuse.Reactive/ObserverMap.uno (hunk 7 above)
Comment (sebbert) [reply Source/Fuse.Reactive/ObserverMap.uno]: I think that sounds like a good approach. Please ping me when it's merged and I'll rebase.

File: Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno (hunk 5 above)
Comment (sebbert) [reply Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: We definitely don't, although it needs to be evaluated in the right context. What's the correct place for this in your opinion? Builtins?

Comment (kusma) [reply Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: TBH, I'm not really sure.

Comment (kusma) [reply Source/Fuse.Reactive.JavaScript/TreeObservable/TreeArray.uno]: Lets just do that as a follow-up to get things moving.

**Comment by kusma** [comment]
Seems like this broke master :(

**Comment by kusma** [comment]
Reverted. Please find the problem and send a new PR.

**Pull request merged**